File: apple/Sources/MeshDropKit/GatewayCommands.swift
import Foundation
import Combine
import OSLog

private let log = Logger(subsystem: "com.welape.meshdrop", category: "GatewayCommands")

/// 把 protocol/companion-bridges.md §1 命令路由到 ShareEngine。
///
/// - dispatch: 接收 JSON 命令字节，路由到 engine，返回 JSON 回执字节。
/// - 事件订阅：listen 一个 connection (closure)，会收到 state_snapshot 首帧 + 后续
///   device_snapshot / pairing_pending / offer_pending / history_snapshot 事件 JSON。
@MainActor
public final class GatewayCommands {
    private let engine: ShareEngine

    public init(engine: ShareEngine) {
        self.engine = engine
    }

    // MARK: - 命令分发

    /// 解析 JSON 命令并执行，返回回执 JSON 字节。
    /// 调用方可以来自任意 actor（内部已 hop 到 MainActor）。
    public func dispatch(_ requestJSON: Data) -> Data {
        guard
            let obj = try? JSONSerialization.jsonObject(with: requestJSON) as? [String: Any]
        else {
            return Self.makeReply(id: "", ok: false, error: "invalid_json")
        }
        let cmdID = (obj["id"] as? String) ?? ""
        let type = (obj["type"] as? String) ?? ""
        let payload = (obj["payload"] as? [String: Any]) ?? [:]

        switch type {
        case "list_devices":
            return Self.makeReply(id: cmdID, ok: true, result: [
                "devices": engine.devices.map(deviceWire),
            ])

        case "get_state":
            return Self.makeReply(id: cmdID, ok: true, result: stateSnapshotPayload())

        case "send_text":
            let peerId = (payload["peerId"] as? String) ?? ""
            let text = (payload["text"] as? String) ?? ""
            guard let dev = engine.devices.first(where: { $0.id == peerId }) else {
                return Self.makeReply(id: cmdID, ok: false, error: "peer_not_found")
            }
            engine.sendText(to: dev, content: text)
            return Self.makeReply(id: cmdID, ok: true)

        case "send_file_ref":
            let peerId = (payload["peerId"] as? String) ?? ""
            let fileRef = (payload["fileRef"] as? String) ?? ""
            guard let dev = engine.devices.first(where: { $0.id == peerId }) else {
                return Self.makeReply(id: cmdID, ok: false, error: "peer_not_found")
            }
            let url = URL(fileURLWithPath: fileRef)
            guard FileManager.default.fileExists(atPath: url.path) else {
                return Self.makeReply(id: cmdID, ok: false, error: "file_ref_invalid")
            }
            engine.sendFile(to: dev, sourceURL: url)
            return Self.makeReply(id: cmdID, ok: true)

        case "accept_offer":
            let offerId = (payload["offerId"] as? String) ?? ""
            guard let uuid = UUID(uuidString: offerId) else {
                return Self.makeReply(id: cmdID, ok: false, error: "bad_offer_id")
            }
            engine.respondToFileOffer(uuid, accept: true)
            return Self.makeReply(id: cmdID, ok: true)

        case "reject_offer":
            let offerId = (payload["offerId"] as? String) ?? ""
            guard let uuid = UUID(uuidString: offerId) else {
                return Self.makeReply(id: cmdID, ok: false, error: "bad_offer_id")
            }
            engine.respondToFileOffer(uuid, accept: false)
            return Self.makeReply(id: cmdID, ok: true)

        case "accept_pairing":
            let pairId = (payload["pairingId"] as? String) ?? ""
            let trust = (payload["trust"] as? Bool) ?? false
            guard let uuid = UUID(uuidString: pairId) else {
                return Self.makeReply(id: cmdID, ok: false, error: "bad_pairing_id")
            }
            engine.respondToPairing(uuid, decision: trust ? .trust : .allowOnce)
            return Self.makeReply(id: cmdID, ok: true)

        case "reject_pairing":
            let pairId = (payload["pairingId"] as? String) ?? ""
            guard let uuid = UUID(uuidString: pairId) else {
                return Self.makeReply(id: cmdID, ok: false, error: "bad_pairing_id")
            }
            engine.respondToPairing(uuid, decision: .reject)
            return Self.makeReply(id: cmdID, ok: true)

        case "clear_history":
            engine.clearHistory()
            return Self.makeReply(id: cmdID, ok: true)

        case "cancel_transfer":
            let itemId = (payload["itemId"] as? String) ?? ""
            guard let uuid = UUID(uuidString: itemId) else {
                return Self.makeReply(id: cmdID, ok: false, error: "bad_item_id")
            }
            engine.cancelTransfer(uuid)
            return Self.makeReply(id: cmdID, ok: true)

        case "delete_history_item":
            let itemId = (payload["itemId"] as? String) ?? ""
            guard let uuid = UUID(uuidString: itemId) else {
                return Self.makeReply(id: cmdID, ok: false, error: "bad_item_id")
            }
            engine.removeHistoryItem(uuid)
            return Self.makeReply(id: cmdID, ok: true)

        default:
            return Self.makeReply(id: cmdID, ok: false, error: "unsupported_command")
        }
    }

    // MARK: - 下载查找

    /// 在 history 中找匹配的 incoming 文件，返回 (filename, fileURL)。
    /// 用于 `GET /api/v1/download/<historyId>`。
    public func fileForHistory(id: UUID) -> (name: String, url: URL)? {
        guard let item = engine.history.first(where: { $0.id == id }),
              item.direction == .incoming,
              case .file(let name, _, let url) = item.kind,
              let url else {
            return nil
        }
        return (name, url)
    }

    // MARK: - 事件订阅

    /// 给一个 WS 连接订阅 engine 状态变化。
    /// 返回的 cancellables 集合一旦释放就停止推送。
    ///
    /// 推送的事件 JSON 与 web client `engine.ts` 期望对齐：
    /// - `state_snapshot` 首帧（含 self / devices / history）
    /// - `device_snapshot`（devices 变化）
    /// - `history_snapshot`（history 变化）
    /// - `pairing_pending`（每个新增的 pairing）
    /// - `offer_pending`（每个新增的 offer）
    public func subscribe(send: @escaping (Data) -> Void) -> Set<AnyCancellable> {
        var bag = Set<AnyCancellable>()

        // 首帧：完整状态快照
        send(Self.encodeEvent(type: "state_snapshot", payload: stateSnapshotPayload()))

        // 后续 device 变化：去重避免初始重复推送
        engine.$devices
            .dropFirst()
            .removeDuplicates()
            .sink { [weak self] devices in
                guard let self else { return }
                let json = Self.encodeEvent(type: "device_snapshot", payload: devices.map(self.deviceWire))
                send(json)
            }
            .store(in: &bag)

        engine.$history
            .dropFirst()
            .removeDuplicates()
            .sink { [weak self] history in
                guard let self else { return }
                let json = Self.encodeEvent(type: "history_snapshot", payload: history.map(self.historyWire))
                send(json)
            }
            .store(in: &bag)

        engine.$pendingPairings
            .dropFirst()
            .scan(([] as [PairingRequest], [] as [PairingRequest])) { state, new in
                (state.1, new)
            }
            .sink { [weak self] previous, current in
                guard let self else { return }
                let newOnes = current.filter { p in !previous.contains(where: { $0.id == p.id }) }
                for p in newOnes {
                    send(Self.encodeEvent(type: "pairing_pending", payload: self.pairingWire(p)))
                }
            }
            .store(in: &bag)

        engine.$pendingFileOffers
            .dropFirst()
            .scan(([] as [PendingFileOffer], [] as [PendingFileOffer])) { state, new in
                (state.1, new)
            }
            .sink { [weak self] previous, current in
                guard let self else { return }
                let newOnes = current.filter { o in !previous.contains(where: { $0.id == o.id }) }
                for o in newOnes {
                    send(Self.encodeEvent(type: "offer_pending", payload: self.offerWire(o)))
                }
            }
            .store(in: &bag)

        // 进行中传输的速率推送：每个 metrics 字典变化，挑出当前 transferring 的
        // history 条目，包成 WireTransferProgress 推给 web。
        engine.$transferMetrics
            .dropFirst()
            .removeDuplicates()
            .sink { [weak self] metrics in
                guard let self else { return }
                for (hid, m) in metrics {
                    guard let item = self.engine.history.first(where: { $0.id == hid }),
                          case let .transferring(done, total) = item.status,
                          case let .file(name, _, _) = item.kind
                    else { continue }
                    var payload: [String: Any] = [
                        "id": hid.uuidString,
                        "peerName": item.peer.name,
                        "fileName": name,
                        "bytesSent": Int(done),
                        "totalBytes": Int(total),
                        "speedBps": Int(m.bytesPerSec.rounded()),
                    ]
                    if let etaSec = m.etaSeconds {
                        payload["etaSeconds"] = etaSec
                    }
                    send(Self.encodeEvent(type: "transfer_progress", payload: payload))
                }
            }
            .store(in: &bag)

        return bag
    }

    // MARK: - schema 编码（与 web/src/lib/engine.ts `Wire*` 对齐）

    private func stateSnapshotPayload() -> [String: Any] {
        return [
            "self": [
                "id": engine.identity.id,
                "displayName": engine.displayName,
                "kind": Self.kindString(for: DeviceOS.current),
                "fingerprint": engine.identity.fingerprint,
            ] as [String: Any],
            "devices": engine.devices.map(deviceWire),
            "history": engine.history.map(historyWire),
            "pendingPairings": engine.pendingPairings.map(pairingWire),
            "pendingOffers": engine.pendingFileOffers.map(offerWire),
        ]
    }

    private func deviceWire(_ d: Device) -> [String: Any] {
        var obj: [String: Any] = [
            "id": d.id,
            "displayName": d.name,
            "kind": Self.kindString(for: d.os),
            "fingerprint": d.fingerprint,
            "online": true,
            "trusted": engine.trusted.contains(where: { $0.fingerprint == d.fingerprint }),
        ]
        if let m = d.model { obj["model"] = m }
        return obj
    }

    private func historyWire(_ h: HistoryItem) -> [String: Any] {
        var obj: [String: Any] = [
            "id": h.id.uuidString,
            "direction": h.direction == .outgoing ? "sent" : "received",
            "peerName": h.peer.name,
            "ok": {
                if case .completed = h.status { return true } else { return false }
            }(),
            "completedAt": Int(h.createdAt.timeIntervalSince1970),
        ]
        switch h.kind {
        case .text(let t):
            obj["kind"] = "text"
            obj["text"] = t
        case .file(let name, let size, _):
            obj["kind"] = "file"
            obj["files"] = [["name": name, "sizeBytes": Int(size)]]
            obj["bytesTransferred"] = Int(size)
        }
        return obj
    }

    private func pairingWire(_ p: PairingRequest) -> [String: Any] {
        return [
            "id": p.id.uuidString,
            "peerName": p.peer.name,
            "fingerprint": p.peer.humanFingerprint,
            "createdAt": Int(p.receivedAt.timeIntervalSince1970),
        ]
    }

    private func offerWire(_ o: PendingFileOffer) -> [String: Any] {
        return [
            "id": o.id.uuidString,
            "peerId": o.peer.id,
            "peerName": o.peer.name,
            "kind": "file",
            "files": [["name": o.fileName, "sizeBytes": Int(o.fileSize)]],
            "createdAt": Int(o.receivedAt.timeIntervalSince1970),
        ]
    }

    private static func kindString(for os: DeviceOS) -> String {
        switch os {
        case .macos: return "mac"
        case .ios: return "ios"
        case .windows: return "win"
        case .android: return "android"
        case .linux: return "linux"
        }
    }

    // MARK: - reply / event 包装

    public nonisolated static func makeReply(id: String, ok: Bool, error: String? = nil, result: Any? = nil) -> Data {
        var obj: [String: Any] = [
            "v": 1,
            "id": id,
            "ok": ok,
        ]
        if let error { obj["error"] = error } else { obj["error"] = NSNull() }
        if let result { obj["result"] = result } else { obj["result"] = NSNull() }
        return (try? JSONSerialization.data(withJSONObject: obj, options: [.fragmentsAllowed])) ?? Data()
    }

    public nonisolated static func encodeEvent(type: String, payload: Any) -> Data {
        let obj: [String: Any] = [
            "v": 1,
            "id": "evt-\(UUID().uuidString.lowercased().prefix(12))",
            "type": type,
            "ts": Int(Date().timeIntervalSince1970),
            "payload": payload,
        ]
        return (try? JSONSerialization.data(withJSONObject: obj, options: [.fragmentsAllowed])) ?? Data()
    }
}

File: windows/MeshDrop/Gateway/GatewayCommands.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;
using MeshDrop.Models;
using MeshDrop.Transport;

namespace MeshDrop.Gateway;

/// <summary>
/// 把 protocol/companion-bridges.md §1 命令路由到 ShareEngine。
/// 输入 JSON 对象，输出回执 JSON 对象。
/// </summary>
internal sealed class GatewayCommands
{
    private readonly ShareEngine _engine;

    public GatewayCommands(ShareEngine engine) { _engine = engine; }

    public string Dispatch(string requestJson)
    {
        Cmd? cmd = null;
        try { cmd = JsonSerializer.Deserialize<Cmd>(requestJson); } catch { }
        if (cmd is null || string.IsNullOrEmpty(cmd.Type))
            return Reply("", false, "invalid_request");

        try
        {
            switch (cmd.Type)
            {
                case "list_devices":
                    return Reply(cmd.Id, true, null, new { devices = _engine.Devices.Select(BridgeDevice.From).ToArray() });
                case "get_state":
                    return Reply(cmd.Id, true, null, new
                    {
                        self = BridgeDevice.From(_engine.SelfDevice),
                        devices = _engine.Devices.Select(BridgeDevice.From).ToArray(),
                        history = _engine.History.Select(BridgeHistory.From).ToArray(),
                        pendingPairings = _engine.PendingPairings.Select(BridgePairing.From).ToArray(),
                        pendingOffers = _engine.PendingFileOffers.Select(BridgeOffer.From).ToArray(),
                    });
                case "send_text":
                    {
                        var p = cmd.Payload.Deserialize<SendTextPayload>() ?? new SendTextPayload();
                        var dev = _engine.Devices.FirstOrDefault(d => d.Id == p.PeerId);
                        if (dev is null) return Reply(cmd.Id, false, "peer_not_found");
                        _engine.SendText(dev, p.Text ?? "");
                        return Reply(cmd.Id, true);
                    }
                case "send_clipboard":
                    {
                        var p = cmd.Payload.Deserialize<SendClipboardPayload>() ?? new SendClipboardPayload();
                        var dev = _engine.Devices.FirstOrDefault(d => d.Id == p.PeerId);
                        if (dev is null) return Reply(cmd.Id, false, "peer_not_found");
                        var content = p.Content ?? "";
                        if (string.IsNullOrEmpty(content)) return Reply(cmd.Id, false, "empty_content");
                        var kind = string.IsNullOrEmpty(p.Kind) ? ShareEngine.ClipKind(content) : p.Kind;
                        _engine.PushClipboard(dev, content, kind);
                        return Reply(cmd.Id, true);
                    }
                case "send_file_ref":
                    {
                        var p = cmd.Payload.Deserialize<SendFileRefPayload>() ?? new SendFileRefPayload();
                        var dev = _engine.Devices.FirstOrDefault(d => d.Id == p.PeerId);
                        if (dev is null) return Reply(cmd.Id, false, "peer_not_found");
                        if (string.IsNullOrEmpty(p.FileRef) || !System.IO.File.Exists(p.FileRef))
                            return Reply(cmd.Id, false, "file_ref_invalid");
                        _engine.SendFile(dev, p.FileRef);
                        return Reply(cmd.Id, true);
                    }
                case "accept_offer":
                    {
                        var p = cmd.Payload.Deserialize<OfferPayload>() ?? new OfferPayload();
                        if (!Guid.TryParse(p.OfferId, out var id)) return Reply(cmd.Id, false, "bad_id");
                        _engine.RespondToFileOffer(id, true);
                        return Reply(cmd.Id, true);
                    }
                case "reject_offer":
                    {
                        var p = cmd.Payload.Deserialize<OfferPayload>() ?? new OfferPayload();
                        if (!Guid.TryParse(p.OfferId, out var id)) return Reply(cmd.Id, false, "bad_id");
                        _engine.RespondToFileOffer(id, false);
                        return Reply(cmd.Id, true);
                    }
                case "accept_pairing":
                    {
                        var p = cmd.Payload.Deserialize<PairingPayload>() ?? new PairingPayload();
                        if (!Guid.TryParse(p.PairingId, out var id)) return Reply(cmd.Id, false, "bad_id");
                        _engine.RespondToPairing(id, p.Trust ? PairingDecision.Trust : PairingDecision.AllowOnce);
                        return Reply(cmd.Id, true);
                    }
                case "reject_pairing":
                    {
                        var p = cmd.Payload.Deserialize<PairingPayload>() ?? new PairingPayload();
                        if (!Guid.TryParse(p.PairingId, out var id)) return Reply(cmd.Id, false, "bad_id");
                        _engine.RespondToPairing(id, PairingDecision.Reject);
                        return Reply(cmd.Id, true);
                    }
                case "clear_history":
                    _engine.ClearHistory();
                    return Reply(cmd.Id, true);
                case "delete_history_item":
                    {
                        var p = cmd.Payload.Deserialize<HistoryItemPayload>() ?? new HistoryItemPayload();
                        if (!Guid.TryParse(p.ItemId, out var id)) return Reply(cmd.Id, false, "bad_id");
                        _engine.RemoveHistoryItem(id);
                        return Reply(cmd.Id, true);
                    }
                default:
                    return Reply(cmd.Id, false, "unknown_command");
            }
        }
        catch (Exception ex)
        {
            return Reply(cmd.Id, false, ex.Message);
        }
    }

    /// <summary>
    /// 把 engine 事件转成 protocol/companion-bridges.md §2 的 event JSON。
    /// </summary>
    public static string EncodeEvent(EngineEvent ev)
    {
        object? payload = ev switch
        {
            EngineEvent.DeviceAdded a => BridgeDevice.From(a.Device),
            EngineEvent.DeviceRemoved r => new { id = r.Id },
            EngineEvent.DeviceUpdated u => BridgeDevice.From(u.Device),
            EngineEvent.PairingPending p => BridgePairing.From(p.Pairing),
            EngineEvent.OfferPending o => BridgeOffer.From(o.Offer),
            EngineEvent.TransferProgress t => new { id = t.Id.ToString(), bytesSent = t.BytesSent, totalBytes = t.TotalBytes, speedBps = t.SpeedBps },
            EngineEvent.TransferDone t => new { id = t.Id.ToString(), ok = t.Ok, error = t.Error },
            EngineEvent.HistoryAdded h => BridgeHistory.From(h.Item),
            EngineEvent.ClipboardReceived c => BridgeClipboard.From(c.Entry),
            _ => null,
        };
        var type = ev switch
        {
            EngineEvent.DeviceAdded => "device_added",
            EngineEvent.DeviceRemoved => "device_removed",
            EngineEvent.DeviceUpdated => "device_updated",
            EngineEvent.PairingPending => "pairing_pending",
            EngineEvent.OfferPending => "offer_pending",
            EngineEvent.TransferProgress => "transfer_progress",
            EngineEvent.TransferDone => "transfer_done",
            EngineEvent.HistoryAdded => "history_added",
            EngineEvent.ClipboardReceived => "clipboard_received",
            _ => "unknown",
        };
        return JsonSerializer.Serialize(new
        {
            v = 1,
            id = "evt-" + Guid.NewGuid().ToString("N").Substring(0, 12),
            type,
            ts = DateTimeOffset.UtcNow.ToUnixTimeSeconds(),
            payload,
        });
    }

    private static string Reply(string id, bool ok, string? error = null, object? result = null)
    {
        return JsonSerializer.Serialize(new
        {
            v = 1,
            id,
            ok,
            error,
            result,
        });
    }

    private sealed class Cmd
    {
        [JsonPropertyName("v")] public int V { get; set; } = 1;
        [JsonPropertyName("id")] public string Id { get; set; } = "";
        [JsonPropertyName("type")] public string Type { get; set; } = "";
        [JsonPropertyName("ts")] public long Ts { get; set; }
        [JsonPropertyName("payload")] public JsonElement Payload { get; set; }
    }

    private sealed class SendTextPayload
    {
        [JsonPropertyName("peerId")] public string PeerId { get; set; } = "";
        [JsonPropertyName("text")] public string Text { get; set; } = "";
    }

    private sealed class SendClipboardPayload
    {
        [JsonPropertyName("peerId")] public string PeerId { get; set; } = "";
        [JsonPropertyName("content")] public string Content { get; set; } = "";
        [JsonPropertyName("kind")] public string Kind { get; set; } = "";
    }

    private sealed class SendFileRefPayload
    {
        [JsonPropertyName("peerId")] public string PeerId { get; set; } = "";
        [JsonPropertyName("fileRef")] public string FileRef { get; set; } = "";
        [JsonPropertyName("name")] public string Name { get; set; } = "";
        [JsonPropertyName("sizeBytes")] public long SizeBytes { get; set; }
        [JsonPropertyName("mime")] public string Mime { get; set; } = "";
    }

    private sealed class OfferPayload
    {
        [JsonPropertyName("offerId")] public string OfferId { get; set; } = "";
    }

    private sealed class PairingPayload
    {
        [JsonPropertyName("pairingId")] public string PairingId { get; set; } = "";
        [JsonPropertyName("trust")] public bool Trust { get; set; }
    }

    private sealed class HistoryItemPayload
    {
        [JsonPropertyName("itemId")] public string ItemId { get; set; } = "";
    }
}

internal sealed record BridgeDevice(string Id, string DisplayName, string Kind, string Model,
                                    string Ip, int RttMs, bool Online, bool Trusted, bool Busy)
{
    public static BridgeDevice From(Device d) => new(
        d.Id, d.Name,
        d.Os switch
        {
            DeviceOS.Macos => "mac",
            DeviceOS.Windows => "win",
            DeviceOS.Ios => "ios",
            DeviceOS.Android => "android",
            DeviceOS.Linux => "linux",
            _ => "linux",
        },
        d.Model ?? "",
        d.Host ?? "",
        0, true, false, false);
}

internal sealed record BridgePairing(string Id, string PeerName, string Fingerprint, long CreatedAt)
{
    public static BridgePairing From(PendingPairing p) => new(
        p.Id.ToString(), p.Peer.Name, p.Peer.HumanFingerprint,
        new DateTimeOffset(p.ReceivedAt).ToUnixTimeSeconds());
}

internal sealed record BridgeOffer(string Id, string PeerId, string PeerName, string Kind,
                                   object[] Files, long CreatedAt)
{
    public static BridgeOffer From(PendingFileOffer o) => new(
        o.Id.ToString(), o.Peer.Id, o.Peer.Name, "file",
        new object[] { new { name = o.FileName, sizeBytes = o.FileSize } },
        new DateTimeOffset(o.ReceivedAt).ToUnixTimeSeconds());
}

internal sealed record BridgeHistory(string Id, string Direction, string PeerName,
                                     string Kind, string? Text, object[]? Files,
                                     long? BytesTransferred, bool Ok, long CompletedAt)
{
    public static BridgeHistory From(HistoryItem h)
    {
        var dir = h.Direction == TransferDirection.Outgoing ? "sent" : "received";
        var ok = h.Status is TransferStatus.Completed;
        var completed = new DateTimeOffset(h.CreatedAt).ToUnixTimeSeconds();
        return h.Kind switch
        {
            HistoryKind.Text t => new(h.Id.ToString(), dir, h.Peer.Name, "text", t.Content, null, null, ok, completed),
            HistoryKind.File f => new(h.Id.ToString(), dir, h.Peer.Name, "file", null,
                new object[] { new { name = f.Name, sizeBytes = f.Size } },
                f.Size, ok, completed),
            _ => new(h.Id.ToString(), dir, h.Peer.Name, "text", null, null, null, ok, completed),
        };
    }
}

internal sealed record BridgeClipboard(string Id, string PeerName, string Kind, string Content, long Ts)
{
    public static BridgeClipboard From(ClipboardEntry e) => new(
        e.Id.ToString(), e.PeerName, e.Kind, e.Content,
        new DateTimeOffset(e.ReceivedAt).ToUnixTimeSeconds());
}
